### ISSUE TITLE
[Phase 2] Sequence Embedding Transformer

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/dynamic-clv-segmentation#3]
## Summary

Encode variable-length transaction sequences into fixed-length embeddings that capture behavioral patterns. These embeddings will serve as additional features for CLV prediction.

**Priority:** 🟡 Medium - Enhances prediction but not strictly required
**Parent Issue:** #1
**Depends On:** Phase 1 (#2)

---

## Objectives

1. Create transformer that converts transaction sequences to fixed-length vectors
2. Capture inter-purchase time patterns (mean, std, trend)
3. Handle edge cases (very few or very many transactions)
4. Maintain sklearn-compatible interface

---

## Files to Create

| File | Lines (est.) | Description |
|------|--------------|-------------|
| `src/features/sequence_encoder.py` | 150 | Sequence embedding transformer |
| `src/features/temporal.py` | 80 | Temporal feature utilities |
| `tests/test_sequence_encoder.py` | 120 | Unit tests |

---

### Temporal Utilities
- [ ] Compute inter-purchase time deltas
- [ ] Calculate trend (linear regression slope)
- [ ] Detect periodicity indicators
- [ ] Handle single-transaction edge case

### Sequence Encoder
- [ ] Define embedding components (16-dim default)
- [ ] Implement padding for short sequences
- [ ] Implement truncation for long sequences
- [ ] Add normalization (fit learns parameters)
- [ ] Ensure sklearn transformer interface

### Testing
- [ ] Test embedding shape consistency
- [ ] Verify inter-purchase time statistics
- [ ] Test edge cases (1 transaction, 1000 transactions)
- [ ] Test with real-world-like sequences

---

## Starter Code

📄 **File:** `src/features/sequence_encoder.py`

```python
"""Encode transaction sequences into fixed-length embeddings."""

from sklearn.base import BaseEstimator, TransformerMixin
import numpy as np
import pandas as pd
from typing import Any


class TransactionSequenceEncoder(BaseEstimator, TransformerMixin):
    """
    Encodes variable-length transaction sequences into fixed-length embeddings.

    Embedding components:
    - Inter-purchase time: mean, std, min, max, trend (5 dims)
    - Monetary value: mean, std, min, max, trend (5 dims)
    - Sequence length (normalized) (1 dim)
    - Recency-weighted activity score (1 dim)
    - Periodicity indicators (4 dims)
    Total: 16 dimensions (default)
    """

    def __init__(
        self,
        embedding_dim: int = 16,
        max_sequence_length: int = 100,
        customer_id_col: str = "customer_id",
        transaction_date_col: str = "transaction_date",
        monetary_col: str = "amount"
    ):
        self.embedding_dim = embedding_dim
        self.max_sequence_length = max_sequence_length
        self.customer_id_col = customer_id_col
        self.transaction_date_col = transaction_date_col
        self.monetary_col = monetary_col

        # Learned normalization parameters
        self._ipt_mean: float | None = None  # inter-purchase time
        self._ipt_std: float | None = None
        self._monetary_mean: float | None = None
        self._monetary_std: float | None = None

    def fit(self, X: pd.DataFrame, y: Any = None) -> "TransactionSequenceEncoder":
        """Learn normalization parameters from training data."""
        # Compute global inter-purchase time statistics
        # Compute global monetary statistics
        # TODO: Implement
        raise NotImplementedError

    def transform(self, X: pd.DataFrame) -> np.ndarray:
        """Transform transaction data to embeddings (one per customer)."""
        # Group by customer, compute embedding for each
        # TODO: Implement
        raise NotImplementedError

    def _encode_sequence(
        self,
        dates: np.ndarray,
        amounts: np.ndarray
    ) -> np.ndarray:
        """Encode a single customer's transaction sequence."""
        # TODO: Implement
        raise NotImplementedError
```

📄 **File:** `src/features/temporal.py`

```python
"""Temporal feature utilities for transaction sequences."""

import numpy as np
from scipy import stats


def compute_inter_purchase_times(dates: np.ndarray) -> np.ndarray:
    """Compute time deltas between consecutive purchases (in days)."""
    if len(dates) < 2:
        return np.array([])
    sorted_dates = np.sort(dates)
    deltas = np.diff(sorted_dates).astype('timedelta64[D]').astype(float)
    return deltas


def compute_trend(values: np.ndarray) -> float:
    """Compute linear trend (slope) of a sequence."""
    if len(values) < 2:
        return 0.0
    x = np.arange(len(values))
    slope, _, _, _, _ = stats.linregress(x, values)
    return slope


def detect_periodicity(
    dates: np.ndarray,
    periods: list[int] = [7, 14, 30, 90]
) -> np.ndarray:
    """Detect periodicity indicators for given periods (days)."""
    if len(dates) < 3:
        return np.zeros(len(periods))

    ipt = compute_inter_purchase_times(dates)
    if len(ipt) == 0:
        return np.zeros(len(periods))

    indicators = []
    for period in periods:
        # Fraction of IPTs close to this period (within 20%)
        tolerance = period * 0.2
        matches = np.abs(ipt - period) < tolerance
        indicators.append(matches.mean())

    return np.array(indicators)
```

---

## Definition of Done

- [ ] Embeddings always have shape `(n_customers, embedding_dim)`
- [ ] Inter-purchase time features computed correctly
- [ ] Edge cases handled gracefully (no NaN/Inf)
- [ ] Sklearn `fit`/`transform` interface works
- [ ] All tests pass

---

## Verification Commands

```bash
pytest tests/test_sequence_encoder.py -v

python -c "
from src.features.sequence_encoder import TransactionSequenceEncoder
import pandas as pd

df = pd.DataFrame({
    'customer_id': ['A']*5 + ['B']*2,
    'transaction_date': pd.date_range('2024-01-01', periods=5).tolist() +
                        pd.date_range('2024-01-15', periods=2).tolist(),
    'amount': [100, 150, 120, 180, 200, 50, 75]
})

encoder = TransactionSequenceEncoder()
embeddings = encoder.fit_transform(df)
print(f'Shape: {embeddings.shape}')  # Expected: (2, 16)
"
```
